actions/sync-calendar/gcal/gcal.go

```diff
@@ -0,0 +1,118 @@
+	svc        *calendar.Service
+}
+
+// NewGCal wraps the calendar specified using the client. CalendarID should be a user's primary calendar (not a shared,
+// calendar), because the Clear() method only works on primary calendars.
+func NewGCal(calendarID string, client *http.Client) (*GCal, error) {
+	if calendarID == "" {
+		return nil, fmt.Errorf("calendar ID cannot be empty")
+	}
+
+	svc, err := calendar.NewService(context.Background(), option.WithHTTPClient(client), option.WithUserAgent(UserAgent))
+	if err != nil {
+		return nil, fmt.Errorf("unable to create Calendar service: %v", err)
+	}
+
+	return &GCal{
+		CalendarID: calendarID,
+		svc:        svc,
+	}, nil
+}
+
+// Client creates an authenticated HTTP Client from the specified service account credentials. The service account
+// must have a Domain-wide Delegation to control user calendars. See
+// https://developers.google.com/admin-sdk/directory/v1/guides/delegation
+func Client(calendarID, jsonCredentialsPath string) (*http.Client, error) {
+	key, err := ioutil.ReadFile(jsonCredentialsPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read JSON credential file: %v", err)
+	}
+	jwtConfig, err := google.JWTConfigFromJSON(key, calendar.CalendarScope)
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate config from JSON credential: %v", err)
+	}
+	// Since apparently service accounts don't have any associated quotas in GSuite,
+	// we must supply a user to charge quota against, and I think they need to have
+	// admin permission on the G Suite account to work.
+	jwtConfig.Subject = calendarID
+
+	return jwtConfig.Client(context.Background()), nil
+}
+
+// Clear all events from the calendar
+func (g *GCal) Clear() error {
+	return g.svc.Calendars.Clear(g.CalendarID).Do()
+}
+
+// Add all events from Schedule S to the calendar.
+func (g *GCal) Schedule(s *schedule.Schedule, stop time.Time) error {
+	if err := s.Validate(); err != nil {
+		return fmt.Errorf("schedule is invalid: %v", err)
+	}
+
+	for i, shift := range s.Shifts {
+		startDate, err := time.Parse(schedule.DateFormat, shift.StartDate)
+		if err != nil {
+			return fmt.Errorf("err parsing shift (index %v) start date (%v): %v", i, shift.StartDate, err)
+		}
+
+		var endDate time.Time
+		if i == len(s.Shifts)-1 {
+			endDate = stop
+		} else {
+			nextShift := s.Shifts[i+1]
+			endDate, err = time.Parse(schedule.DateFormat, nextShift.StartDate)
+			if err != nil {
+				return fmt.Errorf("err parsing next shift (index %v) start date (%v): %v", i, nextShift.StartDate, err)
+			}
+		}
+
+		u := shift.User
+		if shift.UserOverride != "" {
+			u = shift.UserOverride
+		}
+		event := &calendar.Event{
+			Summary: fmt.Sprintf("%v Spinnaker OSS Build Cop", u),
+			Start: &calendar.EventDateTime{
+				Date: startDate.Format(DateFormat),
+			},
+			End: &calendar.EventDateTime{
+				Date: endDate.Format(DateFormat),
+			},
+		}
+
+		_, err = g.svc.Events.Insert(g.CalendarID, event).SendUpdates("none").Do()
```

Comment:
It doesn't look like there is any support in the official Go client libraries. There does appear to be a [user-contributed library that has a Calendar example](https://godoc.org/github.com/jfcote87/google-api-go-client/batch#example-Service--Calendar), but it looks like it goes against the global batch endpoint that is [being turned down](https://developers.googleblog.com/2018/03/discontinuing-support-for-json-rpc-and.html).

FWIW the calendar API limit is [1M queries at day](https://developers.google.com/calendar/pricing) and 500 queries per user over 100s

Reply:
Yeah, it probably makes sense to just do this in a loop rather than move to an endpoint that's in the process of being deprecated.

I guess one thought is whether we should do validation before we start writing to the calendar?  So first prepare all the data, then write it in a loop?  That way at least if there's some error validating/preparing the data we won't clear the calendar and partially update.  I don't know how easy that would be but might be worth it 🤷‍♂

Reply:
Good observation. I've removed the separate `Clear` API and moved it into the Schedule method after all validation, so nothing destructive happens on invalid data.